(PUP-7110) Execute agent run in separate contract

mruzicka opened the pull request:
Before this commit, it was not possible to restart/stop a puppet agent service on Solaris from an agent run executed by that very service without killing the agent run along with the service.
This was because services on Solaris run in so called contracts - a kind of process groups - and during a service restart/stop all processes in the contract are killed. Under normal circumstances the contract is inherited during fork system call so the child process is a member of the same contract as its parent process. This was the case of the puppet run executed by the puppet agent service on Solaris too, and that is why the puppet run could not restart/stop the puppet agent service without killing itself.
So in this commit we simply make sure the agent run is executed in a different contract than its parent (the puppet agent service) which fixes the issue.

puppetcla commented:
CLA signed by all contributors.

MikaelSmith commented on lib/puppet/agent.rb:
```diff
@@ -1,6 +1,167 @@
+class AtFork
+  require 'fiddle'
+
+  raise LoadError, %q(The loaded fiddle version doesn't define 'Fiddle::Handle') unless defined?(Fiddle::Handle)
```
What Fiddle versions don't support this? How likely are we to encounter them in the wild?

If we expect people to run into this error, I'd be interested in making it more graceful (i.e. the AtFork functionality doesn't work but everything else works as before).

MikaelSmith replied:
Oh, I guess the rescue handles old versions of Fiddle or no libcontract by falling back to noop.

joshcooper replied:
If we're using an old version of fiddle, then just `require 'puppet'` will fail, even if we're not on Solaris. Can we move this into an `AtFork` method, so we only check if/when we're on Solaris and need to make a fiddle call?

MikaelSmith replied:
It appears the whole class has a `rescue` block. I haven't worked with that before.

joshcooper replied:
Ah... I think it'd be better to move the raise into the place we try to use it, since ruby will raise a NameError if you try to reference a non-existent constant:

```ruby
begin
  libhandle = Fiddle::Handle.new(library)
rescue NameError => e
  raise _("The loaded fiddle version doesn't define 'Fiddle::Handle'")
rescue Fiddle::DLError => e
  raise LoadError, e.message, e.backtrace
end
```

joshcooper replied:
Ah, sorry I get it now... One concern is that if we are running on Solaris, and we fail to map one of the functions (maybe a newer solaris version changes a method signature?), then we raise a LoadError, catch it, and silently "downgrade" to noop methods. So we could regress on this issue and not know it.

Could we instead define two implementations, one for Solaris, and one for everyone else, something like we do for FileSystem:

```ruby
class AtFork
  @impl = if Facter.value(:operatingsystem) == 'Solaris'
            SolarisAtFork.new
          else
             NoopAtFork.new
          end

  def prepare
    @impl.prepare
  end
end

class NoopAtFork
  def prepare
  end
  ...
end

class SolarisAtFork
  begin
    libhandle = Fiddle::Handle.new(library)
  rescue NameError => e
    raise _("The loaded fiddle version doesn't define 'Fiddle::Handle'")
  rescue Fiddle::DLError => e
    raise LoadError, e.message, e.backtrace
  end
  
  def prepare
    activate_new_contract_template
  end
end
```

mruzicka replied:
A bit of explanation, which hopefully answers some of your questions:
* Without this check travis was failing on ruby 1.9.3 which apparently includes an early version of Fiddle which doesn't define the `Fiddle::Handle` class and does define (and throws) different exceptions (e.g. `DL::DLError` instead of `Fiddle::DLError`) than more modern versions, so in order to avoid having to complicate the implementation to make it compatible with this older versions of fiddle I opted for adding the check.
* The check is really meant to ensure the code is executing with a sufficiently modern version of Fiddle - ideally I'd do something like:
```ruby
raise "Too old version of fiddle." unless Fiddle::Version >= x.y.z
```
and as such it felt reasonable to place it at the beginning.
* I agree that checking for whether the code is executing on Solaris is a more robust approach than the no-op fallback.

mruzicka replied:
A question: What should we do on Solaris when Fiddle is not available or too old?
Should we:
1. Fallback to no-op? Rationale: It's how it used to work and the chances are hight the code won't be restarting the service and thus won't be affected by the problem we are trying to solve.
2. Raise an exception? Rationale: It either works in all scenarios as it supposed to or not at all ...

joshcooper commented on lib/puppet/agent.rb:
```diff
@@ -1,6 +1,167 @@
+
+      raise if ct_tmpl_activate(tmpl_fd) != 0
+    rescue
+      tmpl.close()
```
If someone adds code between the two begin/rescue blocks and that code raises, then close won't be called. Could simplify to:

```ruby
begin
  tmpl = File.open(..)
  begin
    tmpl_fd = tmpl.fileno
    ...
  ensure
    tmpl.close
  end
rescue => detail
  Puppet.error(_("Failed to activate contract template: %{detail}") % { detail: detail }
end
```

mruzicka replied:
I don't think you can avoid that problem (the code above has the same problem if someone adds code between the line with the `File.open` call and the `begin` on the the next line) unless at the expense of more clutter:
```
tmpl = nil
begin
  tmpl = File.open(...)
  ...
rescue
  tmpl.close() unless tmpl.nil?
end
```

joshcooper replied:
True, but I think nested it makes it self-evident that there is a dependency between the open and close. If you have two independent begin/rescues then it's not obvious that they are in fact dependent.

joshcooper commented on lib/puppet/agent.rb:
```diff
@@ -1,6 +1,167 @@
+  end
+
+  def deactivate_contract_template(where)
+    return 0 if @tmpl.nil?
```
Can we just `return if @tmpl.nil?` since the caller doesn't look at the return code?

mruzicka replied:
Absolutely!

joshcooper commented on lib/puppet/agent.rb:
```diff
@@ -1,6 +1,167 @@
+    rescue
+      Puppet.fatal 'Unable to read contract status'
+    ensure
+      stat.close()
```
Similar comment about how `stat.close` might not be called if someone adds code on line 104.

joshcooper commented on lib/puppet/agent.rb:
```diff
@@ -1,6 +1,167 @@
+
+      raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
+      ctid = ct_status_get_id(stathdl)
+      ct_status_free(stathdl)
```
If  line 108 raises, then we never free the `stathdl`. A good way to handle this is to nest begin/rescues:

```ruby
stat = File.open(...)
begin
  stathdl = Fiddle::Pointer.new(0)
  begin
    raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
    ...
  ensure
    ct_status_free(stathdl)
  end
rescue
  ...
ensure
  stat.close
end
```

Or create private helper methods to yield that param and ensure it's closed/released:

```ruby
def with_stat(&block)
  stat = File.open(...)
  yield stat
ensure
  stat.close
end
```

Then the caller can do:

```ruby
with_stat do |stat|
  with_fiddle_pointer do |stathdl|
    raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
    ...
  end
end

mruzicka replied:
If line 108 raises, there is nothing to free. The `stathdl` itself is a ruby managed object and is garbage collected. Only when the call on line 108 succeeds that object points to an opaques structure, which needs to be freed by calling the `ct_status_free`. But as the `ct_status_get_id` on line 109 should not raise - there is no failure mode documented for it: http://docs.oracle.com/cd/E86824_01/html/E54768/ct-status-get-id-3contract.html, the code should always free it on line 110.

joshcooper replied:
Ah right. That said I prefer explicitly managing the pointer lifetimes when interacting with native calls and not relying on garbage collection to deallocate memory, e.g.

```ruby
stathdl = Fiddle::Pointer.new(0)
begin
  raise if ct_status_read(stat.fileno, CTD_COMMON, stathdl.ref) != 0
  begin
    ...
  ensure
    ct_status_free(stathdl)
  end
ensure
  stathdl.free
end
```

FFI has a nice block form that handles that automatically, e.g. https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/api_types.rb#L25-L30, wish Fiddle had something similar.

mruzicka commented:
@joshcooper I'm no sure sure if we really always want to only log the exceptions instead of failing the execution, since for instance leaving the contract template active in the child process would cause every process executed by the child process to run in its own contract, my guess is there is probably quite a few of those (consider facter) and that could potentially lead to exhaustion of some limit on the total number of contracts. So I'd rather fail to surface the problem early.
On the other hand if we fail to active the new contract template it may be desirable to log the exception and continue as in that case the worst we risk is a return to the behavior before this fix.

mruzicka commented on lib/puppet/util/at_fork/solaris.rb:
```diff
@@ -0,0 +1,155 @@
+    end
+
+    begin
+      stathdl = Fiddle::Pointer.new(0)
```
Sorry if I'm stating the obvious, but it seems there is a misunderstanding here:
This does not allocate any bit of memory which would not be managed by ruby. It creates a Fiddle equivalent of something which would in C syntax read:
```c
void *stathdl = NULL;
```
the storage space for the `stathdl` itself is managed by ruby.
The following `ct_status_read` call resets the pointer (note the use `stathdl.ref` - a Fiddle equivalent of the `&stathdl` C syntax) to point to an opaque structure which can be further examined by the  `ct_status_get_*` family of calls and eventually must be freed by the `ct_status_free` call. OTOH there is no need to do anything special about the free the `stathdl` itself as it is garbage collected by ruby automatically when it goes out of scope.

joshcooper replied:
Yeah, makes sense. I assumed that fiddle allocated native memory for the ptr itself, even when the address is 0x0, but it doesn't look like it behaves that way. I assumed it worked like FFI, which does allocate memory:

```ruby
static VALUE
memptr_initialize(int argc, VALUE* argv, VALUE self)
{
    VALUE size = Qnil, count = Qnil, clear = Qnil;
    int nargs = rb_scan_args(argc, argv, "12", &size, &count, &clear);

    memptr_malloc(self, rbffi_type_size(size), nargs > 1 ? NUM2LONG(count) : 1,
        RTEST(clear) || clear == Qnil);
```

About the NULL business, it might be clearer to do `stathdl = Fiddle::NULL`

mruzicka replied:
I'm not sure you if you can reset the Fiddle::NULL pointer constant. But if you can I agree it's a clearer approach, I'll investigate.

mruzicka replied:
It turns out you indeed can reset the `Fiddle::NULL` constant:
```ruby
require 'fiddle'
strtol = Fiddle::Function.new(Fiddle::Handle.new['strtol'], [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_INT], Fiddle::TYPE_LONG)
a = '12xy'
b = Fiddle::NULL
puts "Fiddle::NULL address: #{Fiddle::NULL.to_i}"
puts "Calling strtol function ..."
c = strtol.call(Fiddle::Pointer.to_ptr(a), b.ref, 10)
puts "Fiddle::NULL address: #{Fiddle::NULL.to_i}"
puts "Integer value: #{c}"
puts "Unconverted string: #{b.to_s}"
```
which outputs:
```
Fiddle::NULL address: 0
Calling strtol function ...
Fiddle::NULL address: 140539989261538
Integer value: 12
Unconverted string: xy
```
Which is not something with what to do, so I'll stick with the current implementation.

joshcooper commented on lib/puppet/util/at_fork/solaris.rb:
```diff
@@ -0,0 +1,155 @@
+  def raise_if_error()
+    unless (e = yield) == 0
+      e = SystemCallError.new(nil, e)
+      raise e, e.message, caller
```
A couple of ruby-isms. We don't use parentheses for empty argument lists, so just `def raise_if_error`. However, if a method yields to a block, then you'll want to put that in the argument list, `def raise_if_error(&block)`. Ruby doesn't require it, but it makes the method more readable.

mruzicka replied:
ok

joshcooper commented on lib/puppet/util/at_fork/solaris.rb:
```diff
@@ -0,0 +1,155 @@
+      rescue
+        tmpl.close()
+        raise
+      end
```
You could change that to just:

```ruby
ensure
  tmpl.close
end
```
So that you don't need to re-raise the exception.

mruzicka replied:
No, I can't, because I need the handle to remain open in case no exception is thrown. In that case the handle is stored in an instance variable and later closed in the `deactivate_contract_template` method.

MikaelSmith commented on lib/puppet/util/at_fork.rb:
```diff
@@ -0,0 +1,35 @@
+# - a noop implementation used on all platforms except Solaris (and possibly
+#   even there as a fallback)
+# - a Solaris implementation which ensures the forked process runs in a different
+#   contract than the parent process. This is necessary for agent runs stared by
```
Typo: `stared` should be `started`.

MikaelSmith replied:
This is the only comment I expect a change for. The others were just observations.

MikaelSmith commented on lib/puppet/util/at_fork.rb:
```diff
@@ -0,0 +1,35 @@
+#   killed in the process as a consequence of running in the same contract as the
+#   service.
+module Puppet::Util::AtFork
+  @handler_class = loop do
```
This is a crazy idiom. I suppose it avoids writing and calling a new function just to get simple fall-through and assignment.

mruzicka replied:
Yes, exactly so, I wish there was a method similar to `loop` but which would have called the block only once.
I realize this is not a common use of the `loop` method that's why I added the comments on the break statements used to return values from the construct.

MikaelSmith commented on lib/puppet/util/at_fork.rb:
```diff
@@ -0,0 +1,35 @@
+#   service.
+module Puppet::Util::AtFork
+  @handler_class = loop do
+    if Facter.value(:operatingsystem) == 'Solaris'
```
Same day using Facter here may bite us, but we currently use it in settings parsing so this isn't the first use in the service.

mruzicka replied:
Is there a better way? Perhaps RUBY_PLATFORM?
@joshcooper?

joshcooper replied:
Thor vendored into bundler does that: https://github.com/bundler/bundler/blob/09358d459565c39314eea933020c8f848b3cfce0/lib/bundler/vendor/thor/lib/thor/shell/basic.rb#L378

One issue is it doesn't work in JRuby (but that doesn't apply here). Or maybe [`RbConfig::CONFIG['host_os']`](http://stackoverflow.com/questions/11784109/detecting-operating-systems-in-ruby)?

mruzicka commented on lib/puppet/util/at_fork/solaris.rb:
```diff
@@ -0,0 +1,155 @@
+    begin
+      raise_if_error { ct_tmpl_clear(tmpl.fileno) }
+    rescue => detail
+      Puppet.log_exception(detail, _("Failed to deactivate process contract template in the #{where} process"))
```
@joshcooper I suspect this might cause problems to the i18n's translatable string extraction, effectively those are two distinct messages:
```
Failed to deactivate process contract template in the parent process
```
and
```
Failed to deactivate process contract template in the child process
```
but figuring that out is probably beyond the string extraction ability?
What would be correct approach?

joshcooper replied:
Ah, thanks for catching that. You can do something like:

```ruby
_("Failed to ... %{where} process") % { where: where }
```

mruzicka replied:
But then the translators will only see:
```
Failed to ... %{where} process
```
which may be more difficult to translate than the two distinct non-parametrized phrases.
What I had on my mind was something like:
```ruby
Puppet.log_exception(detail, parent? ? _("... parent process") : _("... child process"))
```
does that make sense?

joshcooper replied:
ah yeah, just the parameter to a boolean and make the two cases constants strings. 👍

mruzicka replied:
k

mruzicka commented:
@joshcooper @MikaelSmith I believe I've addressed all of your comments. The tests are now also included. Note that tests only exercise the functionality of the 'Puppet::Util::AtFork' module, but not of the handlers returned by the module. This is because the handler is either a noop or a very platform specific code which is difficult to test in a generic way beyond verifying it logs the expected messages when some of the system calls fails. I'm happy to add those tests if you feel they provide enough value.

mruzicka commented:
And the tests are passing now too ... yay!

mruzicka commented:
Rebased on latest stable.

MikaelSmith commented:
I'm running this through a quick acceptance suite at https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/ad-hoc/view/vmpooler/job/platform_puppet-agent_intn-van-sys_suite-ad-hoc-vmpooler-puppet-ad-hoc/28/. If it passes, I'll merge.

MikaelSmith commented:
https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/ad-hoc/view/vmpooler/job/platform_puppet-agent_intn-van-sys_suite-ad-hoc-vmpooler-puppet-ad-hoc/28/SLAVE_LABEL=beaker,TEST_TARGET=solaris10-64a/testReport/junit/(root)/tests_direct_puppet/cached_catalog_remediate_local_drift_rb/

A single failure. Maybe not 100% reliable?

mruzicka commented:
The problem is the ruby built by this pipeline (ruby 2.1.9) doesn't include fiddle - search for fiddle in [console output of the packaging build #69](https://jenkins-master-prod-1.delivery.puppetlabs.net/job/platform_puppet-agent_pkg-van-ship_ad-hoc-vmpooler-ad-hoc/BUILD_TARGET=solaris-11-i386,SLAVE_LABEL=beaker/69/consoleFull).
FIddle is only included in ruby 2.3.3 which only seems to by built by the master pipeline.

So the question is: Should I rebase the PR on the master branch or should I rather file a ticket to make sure Fiddle is included in ruby 2.1.9 which we ship with older versions of puppet-agent?

MikaelSmith commented:
I'd discuss it with @McdonaldSeanp maybe, I think he's done a fair amount of the Ruby 2.3 work. If it's not very hard, I'd rather get that into the ruby 2.1.9 line.

mruzicka commented:
Rebased on the master branch.

MikaelSmith commented:
Is the fact that Ruby 2.3 is not default on master going to block merging this ticket?

Magisus commented:
Ruby 2.3 IS currently the default on the master branch...

MikaelSmith commented:
Did I do something wrong to kick it off then? https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/ad-hoc/view/vmpooler/job/platform_puppet-agent_pkg-van-ship_ad-hoc-vmpooler-ad-hoc/BUILD_TARGET=solaris-10-i386,SLAVE_LABEL=beaker/71/consoleFull used 2.1.9.

Magisus commented:
Looks like you might need to rebase your branch, the project config there has 2.1.9, but in master we have 2.3.3: https://github.com/puppetlabs/puppet-agent/blob/master/configs/projects/puppet-agent.rb#L94-L95

mruzicka commented:
Yesterday I kicked off this build: https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/ad-hoc/view/vmpooler/job/platform_puppet-agent_init-van-int_suite-ad-hoc-vmpooler-ad-hoc/87/
And it has completed successfully.
